### PR TITLE
Add WASI support

### PR DESCRIPTION
**Pull request by saghul**
Screen Shot 2024-02-13 at 12 51 18

### CODE REVIEW

**Comment by saghul** [comment]
@bnoordhuis Can you PTAL?

File: CMakeLists.txt [hunk 1]
```diff
@@ -54,6 +54,17 @@ if(MSVC)
     add_compile_definitions(WIN32_LEAN_AND_MEAN)
 endif()
 
+if("${CMAKE_SYSTEM_NAME}" STREQUAL "WASI")
```
Comment (bnoordhuis) [inline CMakeLists.txt]:
```suggestion
if(CMAKE_SYSTEM_NAME STREQUAL "WASI")
```
Or is there a reason for writing it like that?

File: quickjs-libc.c [hunk 2]
```diff
@@ -76,10 +78,12 @@ typedef sig_t sighandler_t;
 extern char **environ;
 #endif
 
+#endif /* _WIN32 */
+
+#if !(defined(_WIN32) || defined(__wasi__))
```
Comment (bnoordhuis) [inline quickjs-libc.c]: I find that hard to read, personally - it's easy to miss the negation. Can I suggest writing it like this?
```suggestion
#if !defined(_WIN32) && !defined(__wasi__)
```

File: quickjs-libc.c [hunk 3]
```diff
@@ -940,6 +947,7 @@ static JSValue js_std_popen(JSContext *ctx, JSValue this_val,
     JS_FreeCString(ctx, mode);
     return JS_EXCEPTION;
 }
+#endif
```
Comment (bnoordhuis) [inline quickjs-libc.c]:
```suggestion
#endif // !defined(__wasi__)
```
Some of those blocks are pretty big and it's not always obvious what if statement an endif belongs to.

**Review by bnoordhuis — APPROVED** [review]

**Comment by bnoordhuis** [comment]
Forgot to say: rad shell prompt! Here's what mine looks like:
```
$
```

**Comment by saghul** [comment]
> Forgot to say: rad shell prompt! Here's what mine looks like:
> 
> ```
> $
> ```

hahaha written in rust no less! https://starship.rs/

**Comment by chqrlie** [comment]
> Forgot to say: rad shell prompt! Here's what mine looks like:
> 
> ```
> $
> ```

With or without a trailing space?

**Comment by bnoordhuis** [comment]
I admit I'm not enough of a purist to forgo that extra blank :)

**Comment by dimisus** [comment]
Awesome!

**Pull request merged**